### ISSUE TITLE
Loading Large Long-Context SFT Dataset leads to RAM OOM

### ISSUE THREAD

**Issue by PranjalChitale** [axolotl-ai-cloud/axolotl#2975]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

The dataset is loaded appropriately and training commences.

### Current behaviour

``[2025-07-24 10:42:41,253] [INFO] [axolotl.utils.data.shared.load_preprocessed_dataset:470] [PID:126508] [RANK:0] Loading prepared dataset from disk at ../[path]``

The dataset loading freezes here, and if we monitor htop, the RAM usage shoots upto almost 1.3T, causing RAM OOMs, resulting in a complete system failure.

### Steps to reproduce

Dataset used : [Dataset](https://huggingface.co/datasets/VarunGumma/updesh-beta-merged)
Size : ~9.5M samples

The dataset is already preprocessed. tokenized and saved to disk and consumes around 552GB on the disk.
The context length of the dataset ranges from ~1K to ~65K tokens.

The training is carried out on 8 nodes each containing 8 MI300X GPUs (Total 64 GPUs with 192 GB VRAM each).
Each machine has ~1.5T of RAM.

### Config yaml

```yaml
datasets:
- path: data.parquet
  type: chat_template
  field_messages: messages
  message_property_mappings:
    role: role
    content: value
train_on_eos: turn
train_on_eot: turn
roles_to_train:
- assistant
dataset_processes: 88
dataset_keep_in_memory: false
dataset_prefetch_factor: 2
dataset_num_workers: 88
dataset_prepared_path: /scratch/last_run_prepared
num_dataset_shards_to_save: 10

sequence_len: 65536
sample_packing: true
pad_to_sequence_len: true
include_tokens_per_second: true
gradient_accumulation_steps: 16
micro_batch_size: 1
num_epochs: 3
group_by_length: false
```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12.10

### axolotl branch-commit

main/9f2bb188a4e21b72802807164f451a311e4d581b

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by VarunGumma** [comment]
Following.

**Comment by alyosha-swamy** [comment]
Seconding this bug

**Comment by NanoCode012** [comment]
Thanks for this report. Let me check with this internally.

> The dataset loading freezes here, and if we monitor htop, the RAM usage shoots upto almost 1.3T, causing RAM OOMs, resulting in a complete system failure.

Is it a sudden jump or progressively? Any details on how it appears? Have you tried decreasing `dataset_processes`? How many vCPUs on each node? What was the RAM usage prior?

Was this a recent change that broke it? How much RAM did it use to tokenize?

---

@alyosha-swamy , could you provide more details on your Issue?

**Comment by PranjalChitale** [comment]
During the tokenization phase, RAM usage stays stable at most 200-250GB. 

The process operates normally until the processed dataset is written to disk.

As soon as dataset loading begins, RAM usage rises progressively and within 1/2 minutes it shoots up to 1.2–1.3TB, triggering an OOM. 

Each node in the cluster has 96 vCPUs (nproc=96), and I am using dataset_processes=64.


If I try loading the processed dataset manually using `load_from_disk`, it loads properly and RAM usage always stays well below 400GB.

**Comment by NanoCode012** [comment]
> If I try loading the processed dataset manually using load_from_disk, it loads properly and RAM usage always stays well below 400GB.

This is weird, as that's all we do too.

https://github.com/axolotl-ai-cloud/axolotl/blob/1d2aa1e4670012c3a4dea62b883abc489629790a/src/axolotl/utils/data/shared.py#L471-L475

When loading with Axolotl, I wonder if it's attempting to load per node. How did you test yours?

**Comment by VarunGumma** [comment]
hey @NanoCode012 , since we see this messages printed per rank, I believe it is loaded N times? 

`[2025-07-24 10:42:41,253] [INFO] [axolotl.utils.data.shared.load_preprocessed_dataset:470] [PID:126508] [RANK:0] Loading prepared dataset from disk at ../[path]`

**Comment by winglian** [comment]
can you try adding this to your config?
```yaml
accelerator_config:
  dispatch_batches: true
```

**Comment by PranjalChitale** [comment]
Tried this, still the same behavior is observed.

**Comment by dipankar** [comment]
Hi @PranjalChitale, I have been working on a solution for this based on Rust based dataset loader. Would you be open to testing this out?

**Comment by alyosha-swamy** [comment]
@dipankar I'd love to check this out

**Comment by dipankar** [comment]
Took a while, but i have created the package **fast-axolotl** 

https://pypi.org/project/fast-axolotl/

Please test, and let me know if it helps.

**Comment by gvinagremartins** [comment]
Are there any updates to this issue? I'm having the same problem

**Comment by NanoCode012** [comment]
> Are there any updates to this issue? I'm having the same problem

No updates unfortunately, we've been occupied with MoE PRs for recent models. If anyone's willing to tackle this, please let us know!

Could you provide more details on the error? Context length, est token size, and RAM available? 

Alternatively, have anyone tried using `streaming: true` to process on demand?

**Comment by gvinagremartins** [comment]
I am trying to train with 32k context length, also tried with 16k context length, with bf16. I am using 8 nodes with 512Gb Ram each. I've tried with `streaming: true` but it was unsustainably slow....
